### PR TITLE
Pre-step for making phpunit green on Magento 2.4

### PR DESCRIPTION
**Pull request by dumega**
# Description
This PR only includes main changes in classes BoltTestCase and TestHelper for the PR https://github.com/BoltApp/bolt-magento2/pull/1067, so it will be more convenient for us to discuss

Fixes: (link Jira ticket)

#changelog Pre-step for making phpunit green on Magento 2.4

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

File: Test/Unit/BoltTestCase.php [hunk 1]
```diff
@@ -3,13 +3,318 @@
...
+use PHPUnit\Framework\Constraint\IsType;
 
-class BoltTestCase extends TestCase
-{
-    protected function skipTestInUnitTestsFlow()
+if (!defined('PHP_CURRENT_VERSION')) {
+    define('PHP_CURRENT_VERSION', (float)phpversion());
+}
+
+if (PHP_CURRENT_VERSION < 7.2) {
```
Comment (dumega) [inline Test/Unit/BoltTestCase.php:16]: Magento 2.2.x supports PHP 7.0.x ~ 7.1.x
and the version of PHPUnit is 6.2.4

Magento 2.3.x supports PHP 7.1.3 ~ 7.3.x
and the version of PHPUnit is 6.5.14

Magento 2.4.x supports PHP 7.4.x
and the version of PHPUnit is 9.1.5

In PHPUnit 9, it adds PHP return type declarations to the functions, but PHP 7.0 has limited supports for return type declarations, especially it does not support `void` as return type, the void return type is introduced in PHP 7.1 . Also the method TestCase::createPartialMock has specific return type \PHPUnit\Framework\MockObject\MockObject, as a result, to make our tests fully work with Magento 2.2.x, I create this condition

File: Test/Unit/BoltTestCase.php [hunk 2]
```diff
@@ -3,13 +3,318 @@
...
+            return $this->getMockBuilder($originalClassName)
+                        ->disableOriginalConstructor()
+                        ->disableOriginalClone()
+                        ->disableArgumentCloning()
+                        ->disallowMockingUnknownTypes()
+                        ->setMethods(empty($methods) ? null : $methods)
+                        ->getMock();
+        }
+        
+        public static function assertAttributeInstanceOf($expected, $attributeName, $classOrObject, $message = '')
```
Comment (dumega) [inline Test/Unit/BoltTestCase.php:89]: Some methods have been deprecated in PHPUniit 9, we have to define them as needed

File: Test/Unit/BoltTestCase.php [hunk 3]
```diff
@@ -3,13 +3,318 @@
...
+            }
+    
+            $this->expectedExceptionMessageRegExp = $messageRegExp;
         }
     }
-}
\ No newline at end of file
+}
+
+if (!class_exists('\PHPUnit\Framework\Constraint\ArraySubset')) {
```
Comment (dumega) [inline Test/Unit/BoltTestCase.php:225]: https://github.com/sebastianbergmann/phpunit/issues/3494
the class assertArraySubset has been deprecate in PHPUnit 9, we have to define it as needed

File: Test/Unit/TestHelper.php [hunk 4]
```diff
@@ -19,7 +19,9 @@ class TestHelper extends TestCase
      */
     public static function getReflectedClass($class)
     {
-        if (is_subclass_of($class, 'PHPUnit_Framework_MockObject_MockObject')) {
+        if (is_subclass_of($class, '\PHPUnit\Framework\MockObject\MockObject')) {
```
Comment (dumega) [inline Test/Unit/TestHelper.php:22]: In PHPUnit 9, the correct class name is \PHPUnit\Framework\MockObject\MockObject

File: Test/Unit/BoltTestCase.php [hunk 5]
```diff
@@ -3,13 +3,318 @@
...
+        {
+
+        }
+        
+        protected function tearDownInternal()
+        {
+
+        }
+        
+        protected function createPartialMock($originalClassName, array $methods): \PHPUnit\Framework\MockObject\MockObject
```
Comment (dumega) [inline Test/Unit/BoltTestCase.php:78]: In PHPUnit 9, createPartialMock would check if the method does exist before creating Mock object, but some methods like `getValue, setValue, getTitle` are not defined as explicit methods, so it causes warning and fail to create Mock object.

File: Test/Unit/BoltTestCase.php (hunk 1 above)
Comment (vitaliyreznikov) [reply Test/Unit/BoltTestCase.php:16]: Could you explain again why you change 7.1 to 7.2?

File: Test/Unit/BoltTestCase.php (hunk 5 above)
Comment (vitaliyreznikov) [reply Test/Unit/BoltTestCase.php:78]: So you assume that if `PHP_CURRENT_VERSION < 7.2` when have PHPUnit 6.
Can we for better readability have this in condition?
Like if `PHP_CURRENT_VERSION < 7.2 OR PHP_UNIT_VERSION < 9.1`

File: Test/Unit/BoltTestCase.php (hunk 1 above)
Comment (dumega) [reply Test/Unit/BoltTestCase.php]: I have changed the conditional statement as you suggest.

File: Test/Unit/BoltTestCase.php (hunk 5 above)
Comment (dumega) [reply Test/Unit/BoltTestCase.php:75]: I have changed the conditional statement as you suggest.

**Review by danac-gs — APPROVED** [review]

**Review by vitaliyreznikov — APPROVED** [review]
Great job Ning!

**Review by adentong — CHANGES_REQUESTED** [review]
Rebase and merge after https://github.com/BoltApp/bolt-magento2/pull/1070

**Comment by adentong** [comment]
also uncomment https://github.com/BoltApp/bolt-magento2/blob/26ae44ccb8d8711e11c4e3a12bb4c33df123a070/.circleci/config.yml#L291

**Pull request closed without merging**